### PR TITLE
[FLINK-9212][flip6] Port SubtasksAllAccumulatorsHandler to new REST endpoint

### PR DESCRIPTION
**Pull request by yew1eb**
## What is the purpose of the change

*Port SubtasksAllAccumulatorsHandler to new REST endpoint*

## Brief change log

  - *Add SubtasksAllAccumulatorsHandler for WebMonitorEndpoint*
  - *Add SubtasksAllAccumulatorsHeaders/SubtasksAllAccumulatorsInfo for rest response*
  - *Add SubtasksAllAccumulatorsInfoTest*

## Verifying this change

*Added test case SubtasksAllAccumulatorsInfoTest to generate the json response*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/SubtasksAllAccumulatorsHandler.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+/**
+ * Request handler for the subtasks all accumulators.
+ */
+public class SubtasksAllAccumulatorsHandler extends AbstractJobVertexHandler<SubtasksAllAccumulatorsInfo, JobVertexMessageParameters> {
+
+	/**
+	 * Instantiates a new subtasks all accumulators handler.
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/SubtasksAllAccumulatorsHandler.java]: we typically don't have javadocs on constructors as they rarely offer any value

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtasksAllAccumulatorsInfo.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import org.apache.flink.runtime.rest.messages.ResponseBody;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonCreator;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Collection;
+import java.util.Objects;
+
+
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtasksAllAccumulatorsInfo.java]: remove empty line

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/SubtasksAllAccumulatorsHandler.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+			subtaskAccumulatorsInfos.add(
+				new SubtasksAllAccumulatorsInfo.SubtaskAccumulatorsInfo(
+					index++,
+					vertex.getCurrentExecutionAttempt().getAttemptNumber(),
+					locationString,
+					userAccumulators
+				));
+		}
+
+		return new SubtasksAllAccumulatorsInfo(vertexId, parallelism, subtaskAccumulatorsInfos);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/SubtasksAllAccumulatorsHandler.java]: the vertex ID should be passed as a `JobVertexID`, not as a string.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtasksAllAccumulatorsInfo.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+ * Response type of the {@link SubtasksAllAccumulatorsHandler}.
+ */
+public class SubtasksAllAccumulatorsInfo implements ResponseBody {
+
+	public static final String FIELD_NAME_ID = "id";
+	public static final String FIELD_NAME_PARALLELISM = "parallelism";
+	public static final String FILED_NMAE_SUBTASKS = "subtasks";
+
+	@JsonProperty(FIELD_NAME_ID)
+	private String id;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtasksAllAccumulatorsInfo.java]: this should be a `JobVertexID`, and a more descriptive field name would be great. (the JSON field name should remain the same ;) )

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/SubtasksAllAccumulatorsHandler.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	@Override
+	protected SubtasksAllAccumulatorsInfo handleRequest(HandlerRequest<EmptyRequestBody, JobVertexMessageParameters> request, AccessExecutionJobVertex jobVertex) throws RestHandlerException {
+		String vertexId = jobVertex.getJobVertexId().toString();
+		int parallelism = jobVertex.getParallelism();
+
+		final List<SubtasksAllAccumulatorsInfo.SubtaskAccumulatorsInfo> subtaskAccumulatorsInfos = new ArrayList<>();
+
+		int index = 0;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/SubtasksAllAccumulatorsHandler.java]: This is not the actual subtask index. Use `vertex.getCurrentExecutionAttempt().getParallelSubtaskIndex()` instead.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/job/SubtasksAllAccumulatorsInfoTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ */
+public class SubtasksAllAccumulatorsInfoTest extends RestResponseMarshallingTestBase<SubtasksAllAccumulatorsInfo> {
+	@Override
+	protected Class<SubtasksAllAccumulatorsInfo> getTestResponseClass() {
+		return SubtasksAllAccumulatorsInfo.class;
+	}
+
+	@Override
+	protected SubtasksAllAccumulatorsInfo getTestResponseInstance() throws Exception {
+		List<SubtasksAllAccumulatorsInfo.SubtaskAccumulatorsInfo> subtaskAccumulatorsInfos = new ArrayList<>();
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/job/SubtasksAllAccumulatorsInfoTest.java]: you now the size of this array already. (3)

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/job/SubtasksAllAccumulatorsInfoTest.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+		List<UserAccumulator> userAccumulators = new ArrayList<>(3);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/job/SubtasksAllAccumulatorsInfoTest.java]: you actually only have 2 accumulators.

**Comment by yew1eb** [comment]
@zentol thanks for your review. I have updated the PR.

**Comment by zentol** [comment]
You have not changed the type of `SubtasksAllAccumulatorsInfo#jobVertexId` to `JobVertexID`.

**Comment by yew1eb** [comment]
i saw legacy/SubtasksAllAccumulatorsHandler.java used gen.writeStringField("id", jobVertex.getJobVertexId().toString());.

if i change the type of `SubtasksAllAccumulatorsInfo#jobVertexId` to `JobVertexID` , 
run the `SubtasksAllAccumulatorsInfoTest` will fail, error log as follow:
> org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "bytes" (class org.apache.flink.runtime.jobgraph.JobVertexID), not marked as ignorable (2 known properties: "lowerPart", "upperPart"])
 at [Source: {"id":{"upperPart":-7327168959427448358,"lowerPart":4189391734336269051,"bytes":"OiO1lgYSKvuaUKqnBA7x2g=="},"parallelism":4,"subtasks":[{"subtask":0,"attempt":0,"host":"host-0","user-accumulators":[{"name":"test name1","type":"test type1","value":"test value1"},{"name":"test name2","type":"test type2","value":"test value2"}]},{"subtask":1,"attempt":1,"host":"host-1","user-accumulators":[{"name":"test name1","type":"test type1","value":"test value1"},{"name":"test name2","type":"test type2","value":"test value2"}]},{"subtask":2,"attempt":2,"host":"host-2","user-accumulators":[{"name":"test name1","type":"test type1","value":"test value1"},{"name":"test name2","type":"test type2","value":"test value2"}]}]}; line: 1, column: 82] (through reference chain: org.apache.flink.runtime.rest.messages.job.SubtasksAllAccumulatorsInfo["id"]->org.apache.flink.runtime.jobgraph.JobVertexID["bytes"])

**Comment by zentol** [comment]
You have to use the `JobVertexIDSerializer` for that field.

**Comment by yew1eb** [comment]
@zentol thanks you. I updated the code. :laughing:

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/SubtasksAllAccumulatorsHandler.java [hunk 8]
```diff
@@ -97,15 +97,14 @@ public static String createSubtasksAccumulatorsJson(AccessExecutionJobVertex job
...
-		int num = 0;
 		for (AccessExecutionVertex vertex : jobVertex.getTaskVertices()) {
 
 			TaskManagerLocation location = vertex.getCurrentAssignedResourceLocation();
 			String locationString = location == null ? "(unassigned)" : location.getHostname();
 
 			gen.writeStartObject();
 
-			gen.writeNumberField("subtask", num++);
+			gen.writeNumberField("subtask", vertex.getCurrentExecutionAttempt().getParallelSubtaskIndex());
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/SubtasksAllAccumulatorsHandler.java]: Let's keep this at it is.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtasksAllAccumulatorsInfo.java [hunk 9]
```diff
@@ -18,41 +18,46 @@
...
 /**
  * Response type of the {@link SubtasksAllAccumulatorsHandler}.
  */
 public class SubtasksAllAccumulatorsInfo implements ResponseBody {
 
-	public static final String FIELD_NAME_ID = "id";
+	public static final String FIELD_NAME_JOB_VERTEX_ID = "id";
 	public static final String FIELD_NAME_PARALLELISM = "parallelism";
-	public static final String FILED_NMAE_SUBTASKS = "subtasks";
+	public static final String FILED_NAME_SUBTASKS = "subtasks";
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtasksAllAccumulatorsInfo.java]: typo: FILED -> FIELD

**Comment by zentol** [comment]
I'll address the remaining issues myself when merging, thanks for working on this!

**Pull request closed without merging**